### ISSUE TITLE
Reject trajectories that generate `NaN`

### ISSUE THREAD

**Issue by jchodera** [openpathsampling/openpathsampling#558]
I'm only as far as repeatedly running into `nan` Exceptions with the bootstrapping scheme of https://github.com/choderalab/ligand-binding-pathsampling/pull/3, but it would be useful to have trajectories with `nan`s automatically rejected or else OpenPathSampling will be almost guaranteed to crash after running for long enough---especially at high temperature.
```
DONE! Completed Bootstrapping cycle step 80 in ensemble 1/5.
Working on Bootstrapping cycle step 81 in ensemble 1/5 .
Working on Bootstrapping cycle step 82 in ensemble 1/5 .
Working on Bootstrapping cycle step 83 in ensemble 1/5 .
Working on Bootstrapping cycle step 84 in ensemble 1/5 .
Working on Bootstrapping cycle step 85 in ensemble 1/5 .
Working on Bootstrapping cycle step 86 in ensemble 1/5 .
Working on Bootstrapping cycle step 87 in ensemble 1/5 .
Working on Bootstrapping cycle step 88 in ensemble 1/5 .
Working on Bootstrapping cycle step 89 in ensemble 1/5 .
Working on Bootstrapping cycle step 90 in ensemble 1/5 .
Working on Bootstrapping cycle step 91 in ensemble 1/5 .
Working on Bootstrapping cycle step 92 in ensemble 1/5 .
Working on Bootstrapping cycle step 93 in ensemble 1/5 .
Traceback (most recent call last):
  File "setup-ops.py", line 168, in <module>
    initial_sample_set = bootstrap.run()
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/openpathsampling/pathsimulator.py", line 456, in run
    bootstrap.run(n_steps_per_round)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/openpathsampling/pathsimulator.py", line 262, in run
    movepath = bootstrapmove.move(self.sample_set)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/openpathsampling/pathmover.py", line 2385, in move
    subchange = self.mover.move(sample_set)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/openpathsampling/pathmover.py", line 1922, in move
    mover.move(sample_set),
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/openpathsampling/pathmover.py", line 2245, in move
    movepath = mover.move(subglobal)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/openpathsampling/pathmover.py", line 1922, in move
    mover.move(sample_set),
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/openpathsampling/pathmover.py", line 1108, in move
    trials = self(*samples)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/openpathsampling/pathmover.py", line 1200, in __call__
    trial_trajectory = self._run(initial_trajectory, shooting_index)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/openpathsampling/pathmover.py", line 1275, in _run
    trajectory, shooting_index
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/openpathsampling/pathmover.py", line 1239, in _make_forward_trajectory
    running=[run_f])
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/openpathsampling/engines/dynamics_engine.py", line 361, in generate
    snapshot = self.generate_next_frame()
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/openpathsampling/engines/openmm/engine.py", line 257, in generate_next_frame
    self.simulation.step(self.n_steps_per_frame)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/simtk/openmm/app/simulation.py", line 132, in step
    self._simulate(endStep=self.currentStep+steps)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/simtk/openmm/app/simulation.py", line 194, in _simulate
    self.integrator.step(10) # Only take 10 steps at a time, to give Python more chances to respond to a control-c.
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/simtk/openmm/openmm.py", line 13564, in step
    return _openmm.LangevinIntegrator_step(self, steps)
Exception: Particle coordinate is nan
```

**Comment by jhprinz** [comment]
Agreed, we should ignore steps where that happens, which is code wise rather simple. 
Question would be how to communicate these problems to the user.

The only thing we need to make sure is that we do not complete a step the produces `NaN`s, otherwise a rerun will continue to fail. I think I made this mistake last time, but just catching failing iterations.

I propose to replace the `.run` with a generator scheme, like I started in the alanine update. This will make these things easier to catch. The idea is that the `.run()` will return an iterator that an be run or even simpler treat the `PathSampling` as an iterator and you can just do next. Running for `n` steps would then correspond to

```
[next(sampler) for i in range(n)]
```

I will make a PR to discuss and solve this.

**Comment by dwhswenson** [comment]
I think this behavior should be engine-dependent and optional. This is the only time I've heard someone say that NaNs shouldn't be a huge red flag that is immediately reported to the user and stops your simulation before you waste more computing time.

**Comment by jhprinz** [comment]
Well, we were talking about this a lot before :) . I mean running high-temperatures is kind of made to fail, so if you use it, you might expect this to happen, while still being fine with the result, if it does not. 

The simplest way would be to just fix the integration step or reduce the temperature. No sure on how to best proceed. We could do the generator stuff and let the user decide what to do.

None the less, we need a quick check for `NaN` in the OpenMMEngine to catch these problem regardless of the way to deal with it.

**Comment by jhprinz** [comment]
> I think this behavior should be engine-dependent and optional.

That could be an easy solution. You can have a changeable option per engine that handles the behavior when `NaN` is detected. I propose these options:

1. `fail` (default) : This will raise an XYZError and will not return a traj
2. `ignore` : Will not do anything (does not make sense really

In addition `retries : int` : Will just retry the full generation and do it again as many times. After that it fails with the given method

While we are at it, we could also properly handle the `trajectory too long` error:
1. `fail` : raise an error
2. `stop` (default) : what we do right now. Note this will affect the chosen ensemble as it effectively cuts of trajectories longer and will lead to a reject.
3. `ignore` : You might want to disable the limit temporarily

I guess better would be to use fail and handle it.

Solution to @jchodera 's problem:

For running bootstrapping we will add a try-catch clause around the generation and redo the step if it fails. In the this special case I would add an option to bootstrapping on what to do when `NaN` appears.

**Comment by dwhswenson** [comment]
> You can have a changeable option per engine that handles the behavior when NaN is detected.

I was thinking something along similar lines. But I think @jchodera wanted to be able to continue path sampling if your path sampling sim raised an error, not just in the bootstrapping. This is like saying that anything with NaN is formally forbidden, so call it (in configuration MC terms) `V(x)=\infty` and therefore `P(x) = 0`: you can make a trial to that, but you must reject it. However, it counts in the average.

You might also want to see the trajectory that caused the error. That could be useful debug information. If you create an `XYZError`, maybe the `e.args` could carry the trajectory, so that it can be saved for debug purposes, but when the code catches an `XYZError` it creates an `ErrorMoveChange`, which is always rejected? Or maybe have the two modes of handling be either `XYZError` or `XYZWarning`?

But broadly speaking, I think the `fail`, `ignore`, `retries` options make sense.

> While we are at it, we could also properly handle the trajectory too long error

This one I really like. Exactly what you describe sounds good to me.

**Comment by jchodera** [comment]
> This is like saying that anything with NaN is formally forbidden, so call it (in configuration MC terms) V(x)=\infty and therefore P(x) = 0: you can make a trial to that, but you must reject it. 

Correct!

> You might also want to see the trajectory that caused the error.

This is unfortunately almost never helpful, I think.

I think the most helpful mode will be the `retries` mode. If something is really wrong, that will ensure that it will stop and raise an exception, but it will let the occasional `nan` not bring the whole thing crashing down.

**Comment by jchodera** [comment]
Just adding that we think https://github.com/choderalab/ligand-binding-pathsampling/pull/3#issuecomment-239531101 is caused by `nan`s as well, even at 300K.

**Comment by jchodera** [comment]
Still running into the `NaN` issue even at 300K:
```
Working on Bootstrapping cycle step 163261 in ensemble 8/49 .
Traceback (most recent call last):
  File "setup-ops.py", line 205, in <module>
    initial_sample_set = bootstrap.run()
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathsimulator.py", line 456, in run
    bootstrap.run(n_steps_per_round)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathsimulator.py", line 262, in run
    movepath = bootstrapmove.move(self.sample_set)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 2385, in move
    subchange = self.mover.move(sample_set)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 1922, in move
    mover.move(sample_set),
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 2245, in move
    movepath = mover.move(subglobal)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 1922, in move
    mover.move(sample_set),
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 1108, in move
    trials = self(*samples)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 1200, in __call__
    trial_trajectory = self._run(initial_trajectory, shooting_index)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 1275, in _run
    trajectory, shooting_index
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 1239, in _make_forward_trajectory
    running=[run_f])
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/engines/dynamics_engine.py", line 361, in generate
    snapshot = self.generate_next_frame()
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/engines/openmm/engine.py", line 257, in generate_next_frame
    self.simulation.step(self.n_steps_per_frame)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/simtk/openmm/app/simulation.py", line 132, in step
    self._simulate(endStep=self.currentStep+steps)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/simtk/openmm/app/simulation.py", line 194, in _simulate
    self.integrator.step(10) # Only take 10 steps at a time, to give Python more chances to respond to a control-c.
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/simtk/openmm/openmm.py", line 4829, in step
    return _openmm.LangevinIntegrator_step(self, steps)
Exception: Particle coordinate is nan
```

**Comment by jchodera** [comment]
I've added much more aggressive equilibration, but even then, I can't seem to get much of anything usable out of explicit solvent simulation due to `nan`s:
```
Working on Bootstrapping cycle step 163261 in ensemble 8/49 .
Traceback (most recent call last):
  File "setup-ops.py", line 205, in <module>
    initial_sample_set = bootstrap.run()
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathsimulator.py", line 456, in run
    bootstrap.run(n_steps_per_round)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathsimulator.py", line 262, in run
    movepath = bootstrapmove.move(self.sample_set)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 2385, in move
    subchange = self.mover.move(sample_set)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 1922, in move
    mover.move(sample_set),
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 2245, in move
    movepath = mover.move(subglobal)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 1922, in move
    mover.move(sample_set),
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 1108, in move
    trials = self(*samples)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 1200, in __call__
    trial_trajectory = self._run(initial_trajectory, shooting_index)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 1275, in _run
    trajectory, shooting_index
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/pathmover.py", line 1239, in _make_forward_trajectory
    running=[run_f])
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/engines/dynamics_engine.py", line 361, in generate
    snapshot = self.generate_next_frame()
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/openpathsampling-0.9.0-py2.7.egg/openpathsampling/engines/openmm/engine.py", line 257, in generate_next_frame
    self.simulation.step(self.n_steps_per_frame)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/simtk/openmm/app/simulation.py", line 132, in step
    self._simulate(endStep=self.currentStep+steps)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/simtk/openmm/app/simulation.py", line 194, in _simulate
    self.integrator.step(10) # Only take 10 steps at a time, to give Python more chances to respond to a control-c.
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/simtk/openmm/openmm.py", line 4829, in step
    return _openmm.LangevinIntegrator_step(self, steps)
Exception: Particle coordinate is nan
```
My guess is we'll need some sort of `nan` protection in ops 1.0 to avoid turning people off of the package even for simple systems.

**Comment by dwhswenson** [comment]
It might also be a problem with the bootstrapping approach. Try using a storage with the bootstrapping and see how the energies evolve with the trajectories. @sonyahanson and I saw something like this (although at that time, there were other problems still in the way the box vectors were handled). The potential energy was really high on the last few trajectories, IIRC.

**Comment by dwhswenson** [comment]
@jchodera : do you fix the random seed? I find it very strange that both of your tracebacks happen at *exactly* the same point:

```
Working on Bootstrapping cycle step 163261 in ensemble 8/49 .
```

**Comment by jchodera** [comment]
Nope, but I might have copied the same batch queue output twice by
accident! Let me double check.

**Comment by dwhswenson** [comment]
Another quick note: if you're getting up to lots of steps/ensemble in bootstrapping (that output is pretty high), then bootstrapping isn't working. Either you're going along a bad uphill path, or the interfaces aren't well defined.

For RETIS, we want the overlap between successive ensembles to be ~20%. This does *not* correspond to 5 steps/ensemble in bootstrapping (the math isn't that simple), but if that number gets really high, something is wrong. I think I've seen as much ~100 steps for an ensemble still give good acceptance ratios, but if you hit 1000 steps/ensemble, I bet there's a problem. Not sure if that counter resets at each ensemble, but even if not, you're around 20k steps/ensemble.

**Comment by jchodera** [comment]
I think we'd like to ditch bootstrapping altogether and just go for SRTIS with automatic weight updating, but we'll have to put the weight updating in when things are stable enough to do so.

**Comment by dwhswenson** [comment]
Agreed, but also the current implementation of SRTIS has its own host of problems (dummy moves; see my rants about how it was ugly in #349 and proof of the problem in #559).

Also, automatic weight updating might be hard with the current storage setup, which is all based on the premise that schemes are immutable. Right now, the bias is part of the mover which is part of the scheme. So you'd have to store an entire new scheme for every step of your simulation. That's a lot of storage overhead. (Maybe there's an easy way around this, but I'm not going to come up with it tonight. There's definitely a hard way around it: create some sort of parameterized scheme, create move changes that store a parameterized version of the mover, not the mover itself, and create a custom simulator.) Note that this is also a challenge when dealing with AMS, so solving both problems at once might be useful.

**Comment by jchodera** [comment]
As a stop-gap solution, we could do batched updates: Collect N samples from the current bias, update the bias and create new movers, repeat. That may work in the short term.

**Comment by jhprinz** [comment]
Isn't this _stop-gap_ idea what Peter did with Weina in their SRTIS paper and seemed to work fine and converge _more robust_ than the original Wang-Landau idea? I would just try this. It seems reasonable with moderate effort

**Comment by dwhswenson** [comment]
> Isn't this stop-gap idea what Peter did with Weina in their SRTIS paper and seemed to work fine and converge more robust than the original Wang-Landau idea?

Sort of. I think @jchodera is thinking about updating based on SAMS, instead of calculating the crossing probability and updating the fixed bias based on that (which is what Weina did).

The goal for 1.0 was to have fixed bias in place, without any updates of the bias. Wrapping a bias updater around that would require a different `PathSimulator`, but that shouldn't be very hard, especially if you want to do batched updates.

**Comment by dwhswenson** [comment]
We hope the original problem of this issue was solved by @jhprinz's PR #570.

**Comment by jhprinz** [comment]
Can this be closed? I think we found the rather stupid bug. Using single precision on OpenCL was the reason. Now it should work.

**Comment by dwhswenson** [comment]
I think the original point of this issue was solved before that, with #570. Of course, it was hard to tell until #573. ;-)

There's some other stuff in here about batched SAMS updating. Probably best to shift that to the relevant separate issue (#524).

**Comment by jchodera** [comment]
> Can this be closed?

Let me give the updated code a spin first...

**Comment by dwhswenson** [comment]
I'm still having the same problem we discussed last night, which we hoped would be solved by #573. After a run of 10k steps, with a clear speed-up in pace (indicating NaNs) after around 5k steps:

```python
>>> step = storage.steps[9000]
>>> step.change.map_tree(lambda x: type(x))
```

```
[openpathsampling.movechange.PathSimulatorMoveChange,
 [openpathsampling.movechange.RandomChoiceMoveChange,
  [openpathsampling.movechange.RandomChoiceMoveChange,
   [openpathsampling.movechange.RandomChoiceMoveChange,
    [openpathsampling.movechange.RejectedNaNSampleMoveChange]]]]]
```

Also, length of the active trajectory, with markers of rejected NaN move changes.

![image](https://cloud.githubusercontent.com/assets/8178546/17936341/c9b2f248-6a1e-11e6-8b7b-4c6a33608d96.png)

**Comment by dwhswenson** [comment]
We saw no evidence of this problem when I first wrote the AD example. I'm going back to earlier code (checking out 28c799111038a6d272b1282fe409980b1e55a4d1, which was the end of #511) to see if it works there.

Before we merged the AD example #482, I ran it for 10k steps of flexible length and 1k steps of fixed length. It seemed to work then. I'll first check if #511 works, since it contains analysis that I need. We may have to do a bit of a (by-hand) bisect to find the problem.

**Comment by jhprinz** [comment]
My first guess was, that we had always the problem but did not recognize it. It is also not clear, if it happens on all platforms. I had problems with `OpenCL` but not on `CUDA`. `CPU` will raise a `NaN` error itself, so I doubt that we had the problem there. 

Would it possible to run it using `CPU` oer night and see if that runs fine? Then we have a definite platform problem. Also, please check, that you run `mixed` precision using

```
cnx = engine.simulation.context
pl = cnx.getPlatform()
{name: pl.getPropertyValue(cnx, name) for name in pl.getPropertyNames()}
```

In my case it runs on the Intel Graphics Card which only supports `single` precision.

**Comment by jhprinz** [comment]
So what can it be? Let me list a few other ideas:

- a weird move that causes the problems? Unlikely because it would not cause `nan`, right?
- something with storage? Unlikely, only thing I can think of is that some saving screws up the numpy arrays. But why now
- Caching? Unlikely, because it does not happen at the same number of frames
- Platform? My best guess so far. Did we change the setup of Alanine?

**Comment by dwhswenson** [comment]
Honestly, I have no idea what the problem is, and I don't see any obvious possibilities. So rather than guess, my approach is the following:

1. I thought I had good results before. Re-run those results, and make sure they actually were good (that I hadn't overlooked errors). I'm doing this now.
2. If the old results were good, use a divide-and-conquer approach to find the PR that broke it. This is basically what `git bisect` would do, but since the test of this takes 6-10 hours, I'll do it by hand, and do it per PR (instead of per commit). Find a PR halfway(ish) between one where things work and where they don't, run the code, and see what comes out.
3. Repeat step 2 until we've narrowed it down to the problematic PR (or maybe a few possibilities). At this point, we can look at what changes were made and see what might be the problem.

However, if I get good results in step 1, I'll probably run the longer fixed TPS simulation before further bug hunting. This is because (a) I *really* need those results for my talk in a week; (b) running these long bug-hunting tests is something that can be easy to do from a conference.

**Comment by jhprinz** [comment]
I also realized that if you create a CUDA engine and you have non CUDA properties, these will cause no properties to be set, which I think is a bug.

If you have, say CUDA properties and initialize an OpenCL engine you get an exception. And `OpenCL` on my MacBookPro cannot use my `Nvidia Card` it only uses `Intel`

**Comment by dwhswenson** [comment]
According to the setup, it is using OpenCL with mixed precision. I don't have CUDA installed (nor do I have an NVIDIA card), so I'm pretty sure trying to run CUDA would just explode.

**Comment by jhprinz** [comment]
@jchodera : The only thing I am currently thinking would be that `state.getPositions(asNumpy=True)` is not enough and we need to run a `state.getPositions(asNumpy=True).copy()`. Could that be possible? `state` is here the openmm.context.getState` object. I remember that way back there was a `.copy()`

**Comment by jhprinz** [comment]
to check for `nan` I use `np.isnan(np.min(snapshot.velocities._value))` which was mentioned as the fastest way to check for `nan`. Could that be a problem? That we find `nan` although there are none?

**Comment by dwhswenson** [comment]
> Could that be possible? state is here the openmm.context.getStateobject. I remember that way back there was a.copy()

Ouch -- that would be expensive. We already spend most of our time (when I profiled) in `getState` and the `getPositions`/`getVelocities`.

**Comment by jhprinz** [comment]
I think that is why we removed it. If it is pulled from the GPU there is no way it can interfere I think.

**Comment by jhprinz** [comment]
I will add some more openmm inspection convenience methods.

**Comment by dwhswenson** [comment]
The run using the old version of the code seemed to work (took a while, due to some memory issues from having two sims open at once, though). @jhprinz and I have found that the NaNs are often associated with:

* high average KE in trials moves (similarly, high max velocity)
* short active path length (in particular, a change in path length distribution)

Note that the occasional NaN wouldn't be a problem if it got rejected and didn't cause problems for sampling (which is what happens when you suddenly have short path lengths).

Here's a script that checks these things, and examples from the code as of 28c7991 (#511) where they look good:

```python
%matplotlib inline
import matplotlib.pyplot as plt
import openpathsampling as paths
import numpy as np

filename = "tps_nc_files/alanine_dipeptide_tps.nc"
storage = paths.Storage(filename, "r")

# only taking every 10th step for simplicity
stride = 10
steps = storage.steps[::stride]
step_nums = [s.mccycle for s in steps]
lengths = [len(s.active[0].trajectory) for s in steps]
engine = storage.engines[0]

def avg_trial_ke(step, engine):
    trial_traj = step.change.trials[0].trajectory  # TPS: only 1
    traj_tot_ke = 0.0
    for snap in trial_traj:
        engine.current_snapshot = snap
        state = engine.simulation.context.getState(getEnergy=True)
        ke = state.getKineticEnergy()
        traj_tot_ke += ke / ke.unit
    return traj_tot_ke / len(trial_traj)
```

Then the plots to check (followed by their results):

```python
# path length by step: should not show significant drift
plt.plot(step_nums, lengths)
plt.xlabel("MC Step")
plt.ylabel("Path length (frames)")
```

![image](https://cloud.githubusercontent.com/assets/8178546/17965971/8edc12ac-6abf-11e6-9a47-881ed1f5107e.png)


```python
# path length distribution: should not show very short (<10 step) peak
plt.hist(lengths, bins=30, normed=True);
plt.xlabel("Path length (frames)")
plt.ylabel("Probability")
```

![image](https://cloud.githubusercontent.com/assets/8178546/17965949/6dbc3034-6abf-11e6-9bf0-ed144e3f6e4d.png)

```python
# kinetic energies of trial by step: when broken, you need a log plot to see it
# run this less frequently because it is much more expensive to calculate
kes = [(s.mccycle, avg_trial_ke(s, engine)) 
       for s in steps[::10] if len(s.change.trials) == 1]
ke_steps, ke_values = zip(*kes)
plt.plot(ke_steps, ke_values)
plt.xlabel("MC Step")
plt.ylabel("Avg KE of trial frames")
```

![image](https://cloud.githubusercontent.com/assets/8178546/17966253/ee7bd584-6ac0-11e6-8e93-2ad97acffc6f.png)


This last one might be showing a little bit of upward drift, which would be worrisome. But so far I have 100% success (of 2 runs) with the older versions of the code, and 0% success (of I think 3 runs) with the new versions of the code. Small sample size, but enough to make me think that maybe we accidentally changed something.

**Comment by dwhswenson** [comment]
To give an idea of what bad versions of some of these plots would look like: for the first plot, compare with https://github.com/openpathsampling/openpathsampling/issues/558#issuecomment-242102266. A bad version of the last plot:

![image](https://cloud.githubusercontent.com/assets/8178546/17966589/a4993b6c-6ac2-11e6-9730-62434bd70297.png)

(note that this one is plotted on a log scale)

**Comment by dwhswenson** [comment]
Actually, looking at those two plots -- is there any reason that the KEs were so low in the newer (incorrect) version plotted on the log scale? 10^2 - 10^3 before it blows up, as compared to 4*10^3 in the older version that worked.

**Comment by jhprinz** [comment]
True, after the first bump it seems to have the correct _temperature_. Did you use the masses as well? Is it possible that we kind of use too low velocities and suddenly there is a rescaling and then it does weird stuff?

**Comment by dwhswenson** [comment]
I'm not sure what you mean by "use the masses". I ran exactly the code in the example notebook.

**Comment by dwhswenson** [comment]
> True, after the first bump it seems to have the correct temperature.

But also, this is where it NaNs. Same simulation as https://github.com/openpathsampling/openpathsampling/issues/558#issuecomment-242102266.

**Comment by jhprinz** [comment]
> I'm not sure what you mean by "use the masses". I ran exactly the code in the example notebook.

When I did that I just computed the squared velocity and assumed mass one. Just to get a number quickly. i did not compute the kinetic energy

**Comment by jhprinz** [comment]
> But also, this is where it NaNs. 

So, my idea is this: We somehow have too little velocities a rescale will fix this, but potential too large values get amplified and the integration step then causes NaN.

**Comment by dwhswenson** [comment]
> When I did that I just computed the squared velocity and assumed mass one. Just to get a number quickly. i did not compute the kinetic energy

I'm getting KE from OpenMM.

> So, my idea is this: We somehow have too little velocities a rescale will fix this, but potential too large values get amplified and the integration step then causes NaN.

Maybe. I'm definitely curious why we're getting such low velocities initially in the newer versions of the code.

It still doesn't make sense: the newer code somehow has very short paths, with velocities that match the correct velocities for longer paths in the older code. You shouldn't have 5-frame paths that connect these states, not with reasonable velocities.

**Comment by jhprinz** [comment]
> not with reasonable velocities.

If you are fast enough you can. But of course it should not happen.

**Comment by jhprinz** [comment]
There is a bump of about a factor of 100, so in velocities a factor of 10. does that help?

**Comment by jhprinz** [comment]
a factor of 1000 could mean a simtk.unit problem. but 10?

**Comment by jchodera** [comment]
> a factor of 1000 could mean a simtk.unit problem. but 10?

Nanometers vs angstroms?

**Comment by dwhswenson** [comment]
Honestly, I don't think that speculation will get us far until we have more information. More information that would be useful:

* if we use OPS to sample trajectories of a reasonable long length -- say 5-10 ps -- what do the temperatures of those trajectories look like? (`FixedLengthTPSNetwork(paths.FullEnsemble(), paths.FullEnsemble(), length=250)` for 5 ps)
* if we test other commits, where do we see the error pop up?

For now, I'm running the long fixed length TPS simulation in the old commit, because I need these results for a talk I'm giving next week. The sim will take until tomorrow, and then the analysis will take through Saturday. (And I need to have a back-up plan for the talk if the sim fails).

**Comment by dwhswenson** [comment]
Quick update on this: the old version (circa July 7) runs fine. (And I spent the weekend putting the results into a talk.) I'm about to start trying some bisect-style analysis. Worked on July 7; fails as of late August. For my bisect, I'll start on August 1; c4f219886fe0519121aea5aa4f3751b42c1271f0, merge of #500.

(Will update again later once I've reduced the range of possible error-causing commits.)

**Comment by dwhswenson** [comment]
Playing with it a bit more, it looks like pure Python fails for the older code (the July 7 #511 version), although it works fine in ipynb. Not sure what the difference is there, but it means that the error is was not recently introduced (maybe the thing that protected the ipynb was removed, though).

The only thing I can think is that it is a problem with the integrator. I think we're get too big of a bump in velocity. We don't get these problems with the velocity-rescale integrator in Gromacs [Bussi, Donadio, Parrinello version](http://scitation.aip.org/content/aip/journal/jcp/126/1/10.1063/1.2408420). @jchodera : do you think VVVR might fix it (we can get VVVR from `openmmtools`, right?)

**Comment by jchodera** [comment]
The OpenMM integrator you were using at high temperature is:
```
hi_T_integrator = mm.LangevinIntegrator(
    500*unit.kelvin, 
    1.0/unit.picoseconds, 
    2.0*unit.femtoseconds)
hi_T_integrator.setConstraintTolerance(0.00001)
```
The `1/ps` collision rate is very big, and `2 fs` is pretty large for a timestep. This is a leapfrog Langevin integrator from Chris Sweet and Jesus Izaguirre, so you presumably have to integrate out an extra step and flip the velocity sign whenever you want to reverse the trajectory in order to have the appropriate half-step time lag between positions and velocities. You wouldn't have to do that with VVVR or any other velocity Verlet formulation integrator where the positions and timestep are synchronized.

Regarding why IPython might be different from Python: Are you sure the same `Platform` is being used? I could envision paths being set up a bit differently where IPython may load different `Platform`s, which could make different precision models available.

**Comment by dwhswenson** [comment]
> The `1/ps` collision rate is very big

I thought we'd previously said that the collision rate was moderate to small for bio systems? What's the trade-off here? I thought that a larger collision frequency would help quench any extreme velocities.

> `2 fs` is pretty large for a timestep

We used 2 fs on AD with the Gromacs vel-rescale calculation, so I'm surprised if that is a problem.

> This is a leapfrog Langevin integrator

Ahh... this could be a problem. We're not doing anything fancy like that, and I seem to remember @bolhuis saying that leapfrog was not just *formally* bad, but also caused practical problems. (Don't remember details, though.)

> You wouldn't have to do that with VVVR or any other velocity Verlet formulation integrator where the positions and timestep are synchronized.

I think that's the thing to do, then. We should use the VVVR from openmmtools?

> Regarding why IPython might be different from Python: Are you sure the same Platform is being used?

Don't know -- that would be my guess (or different platform options getting passed). By performance, it can't be running CPU, but it might be different precision treatment in OpenCL. (I can't check it this week -- at a conference in Leiden; can't reproduce the IPython-in-Jupyter-notebook-on-work-desktop environment from here.)

**Comment by jchodera** [comment]
Whoops, 1/ps is SMALL. Not sure why I wrote big.

Maybe try VVVR from openmmtools if you are already using it as a
dependency?

**Comment by dwhswenson** [comment]
I'll try the openmmtools VVVR next, and maybe bump the collision rate up the ~2.0-5.0 as well -- thanks for the help!

**Comment by dwhswenson** [comment]
Tried it with the VVVR integrator overnight, using pure-Python script (i.e., not in an ipynb). That is the version that had consistently had NaN problems. It looks like VVVR is promising as a fix.

Ran with collision rate at `1.0/ps` and time step at `2.0 fs`. Average MC steps/second over the last 100 steps was just barely faster than total average (0.44/sec compared to 0.39/sec). No errors.

Plots -- Path length at every (10th) MC step:

![image](https://cloud.githubusercontent.com/assets/8178546/18231357/220b3d3c-72b8-11e6-9bf4-bf66b602bb76.png)

No evidence of getting stuck in super-short paths (this one even managed to sample one path that was longer than I've seen before in this system!)

Average trial KE at every (100th) MC step:

![image](https://cloud.githubusercontent.com/assets/8178546/18231366/88509ace-72b8-11e6-8366-d45d083e8b96.png)

No evidence of sudden jumps, nor occasional trajectories with orders of magnitude more KE.

In other words: ***Using VVVR might fix the problem.***

@jchodera : could you switch to VVVR in https://github.com/choderalab/ligand-binding-pathsampling/pull/4 (as well as updating to the most recent OPS) and see if that fixes your problems there?

I'm also going to re-run a couple times here; if it works for @jchodera and if I can't reproduce the error again, then I think we can say using VVVR is the solution.

**Comment by dwhswenson** [comment]
A couple more little details from that run:

```python
import collections
collections.Counter([type(step.change.subchange.subchange.subchange.subchange)
                     for step in steps[1::]])
```

(where `steps` is actually every 10th MC step)

```
Counter({openpathsampling.movechange.AcceptedSampleMoveChange: 574,
         openpathsampling.movechange.RejectedSampleMoveChange: 425,
         openpathsampling.movechange.RejectedNaNSampleMoveChange: 1})
```

So we *do* still get a `NaN`, but it doesn't seem to destroy the simulation. However, I'll check when it happens:

```python
for step in steps[1::]:
    subch = step.change.subchange.subchange.subchange.subchange
    target_type = paths.RejectedNaNSampleMoveChange
    if isinstance(subch, target_type):
        print step.mccycle
```

```
9470
```

This may be a bit of a concern, because we would often see a few more solitary `NaN`s before the simulation actually blew up (and this one comes near the end of the simulation). So I don't want to promise that VVVR fixes everything, but I'm still hopeful.

**Comment by bolhuis** [comment]
Just to kick in a few open doors: 
1. Velocity verlet is the natural integrator for path sampling. If one uses leap frog, one has to reverse the momenta and do a half time step, velocity only integration. This is how we originally implemented tps, in the jarek scripts. VV does not gave this problem, so if we can set this to the standard setting it is great.
2.  Velocity rescaling should be okay, but a very big change in velocity could still cause havoc.
3. Could the problem lie in the short paths, and the lack of proper relaxation? In other words, a gradual increment of the velocity each time a shot has been accepted?
(You probably already excluded this, not?)

**Comment by dwhswenson** [comment]
Quick replies to @bolhuis, then a summary of recent results:

> Velocity verlet is the natural integrator for path sampling

Yeah, I didn't realize we had a leapfrog-based integrator. Managing that extra half-step could be done by the engine, but that would mean the engines would have to know to handle VV differently from leapfrog. I'd rather just say to users, "Use Velocity Verlet" (at least, for now).

> Could the problem lie in the short paths, and the lack of proper relaxation? In other words, a gradual increment of the velocity each time a shot has been accepted?

I'm not entirely sure what you mean here -- at least, I don't see why a gradual increase wouldn't be offset by a similarly probable gradual decrease. But we were definitely seeing a problem due to short paths: in https://github.com/openpathsampling/openpathsampling/issues/558#issuecomment-242102266 you can see that the path length gets very short before we start getting a lot of moves rejected because of NaN.

***

In any case, I think that VVVR might be enough to fix it. I'm still running with `2.0 fs` time step and only `1.0/ps` collision rate. I've now run `3 * 10k` steps, and the simulations finish without complaint. I *do* see some NaN-rejected moves, but they make up 0.1-0.2% of moves, and are distributed rather evenly throughout the simulation. Importantly, they don't cause the "NaN hole" we were falling into in 100% of simulations before. I expect that increasing the collision frequency or decreasing the time step could reduce the NaNs further. (All of this is about flex TPS on AD.)

I'll make this switch in the AD examples as part of #571. @jchodera, the remaining question is whether your problems in the ligand unbinding example are fixed by VVVR (plus a post-#570 version of OPS). If that works, then I think we can close this issue.

**Comment by jchodera** [comment]
Woohoo! I'll try the ligand unbinding simulations with VVVR.

**Comment by jhprinz** [comment]
I think we agreed that this is solved. Can we close this one then?

**Comment by jchodera** [comment]
Thanks! I have had no problems with the various updates while using velocity Verlet formulation integrators.
